File: scripts/models/hf_model_card_extractor.py
#!/usr/bin/env python3
"""
HuggingFace Model Card Extractor and Mapper
Extracts model card metadata and maps to orchestration task categories
"""

import json
import sys
from pathlib import Path
from typing import Dict, List, Optional, Any
import requests

# Mapping from HuggingFace tags/pipeline_tag to our orchestration categories
HF_TO_ORCHESTRATION_MAPPING = {
    # Math & Reasoning
    "math": ["math"],
    "reasoning": ["reasoning", "math"],
    "question-answering": ["reasoning"],
    
    # Code
    "text-generation": ["code"],  # Many text-gen models do code
    "code": ["code"],
    "code-generation": ["code"],
    
    # Summarization & Document Processing
    "summarization": ["summarization"],
    "document-question-answering": ["ocr_extraction", "reasoning"],
    "document": ["ocr_extraction"],
    
    # Translation (important for Arabic)
    "translation": ["relational"],  # Cross-lingual relationships
    "multilingual": ["relational"],
    
    # Time Series
    "time-series": ["time_series"],
    "forecasting": ["time_series"],
    
    # Vector Search & Embeddings
    "feature-extraction": ["vector_search"],
    "sentence-similarity": ["vector_search"],
    "embeddings": ["vector_search"],
    
    # Graph & Relational
    "table-question-answering": ["relational", "graph"],
    "tabular": ["relational"],
    
    # OCR & Extraction
    "image-to-text": ["ocr_extraction"],
    "ocr": ["ocr_extraction"],
}

# Infer categories from model name/description
MODEL_NAME_PATTERNS = {
    "math": ["gsm", "math", "calc"],
    "code": ["code", "coder", "starcoder", "codegen"],
    "reasoning": ["reason", "think", "chain-of-thought", "cot"],
    "summarization": ["summar", "abstract"],
    "time_series": ["forecast", "timeseries", "time-series"],
    "relational": ["sql", "table", "tabular"],
    "graph": ["graph", "cypher", "neo4j"],
    "vector_search": ["embed", "retriev", "rag"],
    "ocr_extraction": ["ocr", "document", "vision", "extract"],
}

# Language-specific enhancements
LANGUAGE_MAPPINGS = {
    "ar": ["relational"],  # Arabic models benefit from cross-lingual understanding
    "multilingual": ["relational"],
}


def fetch_model_card(repo_id: str) -> Optional[Dict[str, Any]]:
    """Fetch model card from HuggingFace API"""
    try:
        url = f"https://huggingface.co/api/models/{repo_id}"
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"Warning: Could not fetch model card for {repo_id}: {e}", file=sys.stderr)
        return None


def map_to_orchestration_categories(model_data: Dict[str, Any]) -> List[str]:
    """Map HuggingFace model metadata to orchestration categories"""
    categories = set()
    
    # 1. Map from pipeline_tag
    pipeline_tag = model_data.get("pipeline_tag", "")
    if pipeline_tag in HF_TO_ORCHESTRATION_MAPPING:
        categories.update(HF_TO_ORCHESTRATION_MAPPING[pipeline_tag])
    
    # 2. Map from tags
    tags = model_data.get("tags", [])
    for tag in tags:
        if tag in HF_TO_ORCHESTRATION_MAPPING:
            categories.update(HF_TO_ORCHESTRATION_MAPPING[tag])
    
    # 3. Infer from model name
    model_id = model_data.get("modelId", "").lower()
    for category, patterns in MODEL_NAME_PATTERNS.items():
        if any(pattern in model_id for pattern in patterns):
            categories.add(category)
    
    # 4. Check language tags
    for tag in tags:
        if tag.startswith("language:"):
            lang = tag.split(":")[1]
            if lang in LANGUAGE_MAPPINGS:
                categories.update(LANGUAGE_MAPPINGS[lang])
    
    # 5. Check for multilingual
    if any("multilingual" in tag for tag in tags):
        categories.update(LANGUAGE_MAPPINGS["multilingual"])
    
    return sorted(list(categories))


def extract_model_metadata(model_data: Dict[str, Any]) -> Dict[str, Any]:
    """Extract relevant metadata from HuggingFace model card"""
    
    metadata = {
        "model_id": model_data.get("modelId", ""),
        "name": model_data.get("modelId", "").split("/")[-1],
        "author": model_data.get("author", ""),
        "pipeline_tag": model_data.get("pipeline_tag", ""),
        "tags": model_data.get("tags", []),
        "downloads": model_data.get("downloads", 0),
        "likes": model_data.get("likes", 0),
        "library_name": model_data.get("library_name", ""),
        "license": model_data.get("license", ""),
        "languages": [],
        "datasets": [],
        "metrics": {},
    }
    
    # Extract languages
    for tag in metadata["tags"]:
        if tag.startswith("language:"):
            metadata["languages"].append(tag.split(":")[1])
        elif tag.startswith("dataset:"):
            metadata["datasets"].append(tag.split(":")[1])
    
    # Extract card data if available
    if "cardData" in model_data:
        card = model_data["cardData"]
        metadata["base_model"] = card.get("base_model", "")
        metadata["metrics"] = card.get("model-index", [{}])[0].get("results", [])
    
    # Map to orchestration categories
    metadata["orchestration_categories"] = map_to_orchestration_categories(model_data)
    
    # Determine agent types
    metadata["agent_types"] = determine_agent_types(metadata["orchestration_categories"])
    
    # Extract benchmark info if available
    metadata["benchmarks"] = extract_benchmarks(model_data)
    
    return metadata


def determine_agent_types(categories: List[str]) -> List[str]:
    """Determine which agent types this model supports"""
    agent_types = set()
    
    # Inference agents (all models)
    agent_types.add("inference")
    
    # Tool agents (for specific categories)
    tool_categories = {"vector_search", "ocr_extraction", "relational", "graph"}
    if any(cat in tool_categories for cat in categories):
        agent_types.add("tool")
    
    # Orchestrator agents (for reasoning/summarization)
    orchestrator_categories = {"reasoning", "summarization"}
    if any(cat in orchestrator_categories for cat in categories):
        agent_types.add("orchestrator")
    
    return sorted(list(agent_types))


def extract_benchmarks(model_data: Dict[str, Any]) -> Dict[str, Any]:
    """Extract benchmark results from model card"""
    benchmarks = {}
    
    if "cardData" in model_data and "model-index" in model_data["cardData"]:
        for entry in model_data["cardData"]["model-index"]:
            if "results" in entry:
                for result in entry["results"]:
                    dataset = result.get("dataset", {}).get("name", "unknown")
                    metrics = result.get("metrics", [])
                    if metrics:
                        benchmarks[dataset] = {
                            metric["name"]: metric["value"]
                            for metric in metrics
                        }
    
    return benchmarks


def enrich_model_registry(registry_path: str, model_entries: List[Dict[str, Any]]) -> None:
    """Enrich MODEL_REGISTRY.json with HuggingFace model card data"""
    
    # Load existing registry
    with open(registry_path, 'r') as f:
        registry = json.load(f)
    
    # Enrich each model
    for i, model in enumerate(registry["models"]):
        hf_repo = model.get("hf_repo")
        if not hf_repo:
            continue
        
        print(f"Fetching model card for: {hf_repo}")
        model_data = fetch_model_card(hf_repo)
        
        if model_data:
            metadata = extract_model_metadata(model_data)
            
            # Add orchestration metadata
            model["orchestration_categories"] = metadata["orchestration_categories"]
            model["agent_types"] = metadata["agent_types"]
            model["benchmarks"] = metadata["benchmarks"]
            model["hf_metadata"] = {
                "downloads": metadata["downloads"],
                "likes": metadata["likes"],
                "license": metadata["license"],
                "languages": metadata["languages"],
                "datasets": metadata["datasets"],
                "pipeline_tag": metadata["pipeline_tag"],
            }
            
            print(f"  ✓ Categories: {', '.join(metadata['orchestration_categories'])}")
            print(f"  ✓ Agent Types: {', '.join(metadata['agent_types'])}")
            if metadata["benchmarks"]:
                print(f"  ✓ Benchmarks: {list(metadata['benchmarks'].keys())}")
        else:
            print(f"  ✗ Could not fetch model card")
    
    # Save enriched registry
    with open(registry_path, 'w') as f:
        json.dump(registry, f, indent=2)
    
    print(f"\n✓ Registry enriched and saved to {registry_path}")


def main():
    if len(sys.argv) < 2:
        print("Usage: python hf_model_card_extractor.py <path_to_MODEL_REGISTRY.json>")
        print("\nOr to test a single model:")
        print("  python hf_model_card_extractor.py --test <hf_repo_id>")
        sys.exit(1)
    
    if sys.argv[1] == "--test":
        # Test mode - fetch and display single model
        if len(sys.argv) < 3:
            print("Error: Please provide HuggingFace repo ID")
            sys.exit(1)
        
        repo_id = sys.argv[2]
        print(f"Fetching model card for: {repo_id}\n")
        
        model_data = fetch_model_card(repo_id)
        if model_data:
            metadata = extract_model_metadata(model_data)
            print(json.dumps(metadata, indent=2))
        else:
            print("Failed to fetch model card")
            sys.exit(1)
    else:
        # Enrich registry mode
        registry_path = sys.argv[1]
        
        if not Path(registry_path).exists():
            print(f"Error: Registry file not found: {registry_path}")
            sys.exit(1)
        
        print("Enriching MODEL_REGISTRY.json with HuggingFace model card data...")
        print("This may take a minute...\n")
        
        enrich_model_registry(registry_path, [])
        
        print("\n✅ Model registry enriched successfully!")
        print("\nThe registry now includes:")
        print("  • Orchestration categories (math, code, reasoning, etc.)")
        print("  • Agent types (inference, tool, orchestrator)")
        print("  • Benchmark results from HuggingFace")
        print("  • Language and dataset information")
        print("  • Download counts and popularity metrics")


if __name__ == "__main__":
    main()

File: src/serviceCore/nLocalModels/orchestration/hf_model_card_extractor.zig
//! HuggingFace Model Card Extractor
//! Extracts model specifications, benchmarks, and metadata from HuggingFace model cards
//!
//! Features:
//! - Parse model card README for specifications and benchmarks
//! - Extract structured metadata from HF API
//! - Map pipeline tags to orchestration categories
//! - Download and validate benchmark scores
//! - Support batch processing of MODEL_REGISTRY.json

const std = @import("std");
const Allocator = std.mem.Allocator;

/// Orchestration category mappings
const PIPELINE_TAG_MAPPING = std.ComptimeStringMap([]const []const u8, .{
    .{ "math", &[_][]const u8{"math"} },
    .{ "reasoning", &[_][]const u8{ "reasoning", "math" } },
    .{ "question-answering", &[_][]const u8{"reasoning"} },
    .{ "text-generation", &[_][]const u8{"code"} },
    .{ "code", &[_][]const u8{"code"} },
    .{ "code-generation", &[_][]const u8{"code"} },
    .{ "summarization", &[_][]const u8{"summarization"} },
    .{ "translation", &[_][]const u8{"relational"} },
    .{ "multilingual", &[_][]const u8{"relational"} },
    .{ "feature-extraction", &[_][]const u8{"vector_search"} },
    .{ "sentence-similarity", &[_][]const u8{"vector_search"} },
});

const NAME_PATTERNS = std.ComptimeStringMap([]const []const u8, .{
    .{ "math", &[_][]const u8{ "gsm", "math", "calc" } },
    .{ "code", &[_][]const u8{ "code", "coder", "starcoder", "codegen" } },
    .{ "reasoning", &[_][]const u8{ "reason", "think", "chain-of-thought", "cot" } },
    .{ "vector_search", &[_][]const u8{ "embed", "retriev", "rag" } },
});

pub const ModelInfo = struct {
    allocator: Allocator,
    hf_metadata: std.StringHashMap(std.json.Value),
    specifications: std.StringHashMap([]const u8),
    benchmarks: std.StringHashMap(BenchmarkScore),
    hardware: std.StringHashMap([]const u8),
    orchestration_categories: std.ArrayList([]const u8),
    agent_types: std.ArrayList([]const u8),
    
    pub const BenchmarkScore = struct {
        score: f64,
        date: []const u8,
    };
    
    pub fn init(allocator: Allocator) ModelInfo {
        return .{
            .allocator = allocator,
            .hf_metadata = std.StringHashMap(std.json.Value).init(allocator),
            .specifications = std.StringHashMap([]const u8).init(allocator),
            .benchmarks = std.StringHashMap(BenchmarkScore).init(allocator),
            .hardware = std.StringHashMap([]const u8).init(allocator),
            .orchestration_categories = std.ArrayList([]const u8){},
            .agent_types = std.ArrayList([]const u8){},
        };
    }
    
    pub fn deinit(self: *ModelInfo) void {
        self.hf_metadata.deinit();
        
        var spec_iter = self.specifications.iterator();
        while (spec_iter.next()) |entry| {
            self.specifications.allocator.free(entry.key_ptr.*);
            self.specifications.allocator.free(entry.value_ptr.*);
        }
        self.specifications.deinit();
        
        var bench_iter = self.benchmarks.iterator();
        while (bench_iter.next()) |entry| {
            self.benchmarks.allocator.free(entry.key_ptr.*);
            self.benchmarks.allocator.free(entry.value_ptr.date);
        }
        self.benchmarks.deinit();
        
        var hw_iter = self.hardware.iterator();
        while (hw_iter.next()) |entry| {
            self.hardware.allocator.free(entry.key_ptr.*);
            self.hardware.allocator.free(entry.value_ptr.*);
        }
        self.hardware.deinit();
        
        for (self.orchestration_categories.items) |cat| {
            self.allocator.free(cat);
        }
        self.orchestration_categories.deinit();
        
        for (self.agent_types.items) |agent| {
            self.allocator.free(agent);
        }
        self.agent_types.deinit();
    }
};

pub const HFModelCardExtractor = struct {
    allocator: Allocator,
    verbose: bool,
    http_client: std.http.Client,
    
    pub fn init(allocator: Allocator, verbose: bool) HFModelCardExtractor {
        return .{
            .allocator = allocator,
            .verbose = verbose,
            .http_client = std.http.Client{ .allocator = allocator },
        };
    }
    
    pub fn deinit(self: *HFModelCardExtractor) void {
        self.http_client.deinit();
    }
    
    fn log(self: *HFModelCardExtractor, comptime fmt: []const u8, args: anytype) void {
        if (self.verbose) {
            std.debug.print("[INFO] " ++ fmt ++ "\n", args);
        }
    }
    
    /// Extract comprehensive model information from HuggingFace
    pub fn extractModelInfo(self: *HFModelCardExtractor, hf_repo: []const u8) !ModelInfo {
        self.log("Extracting info for: {s}", .{hf_repo});
        
        var info = ModelInfo.init(self.allocator);
        errdefer info.deinit();
        
        // Fetch API info
        if (self.fetchAPIInfo(hf_repo)) |api_data| {
            defer self.allocator.free(api_data);
            try self.parseAPIInfo(&info, api_data);
        } else |err| {
            std.debug.print("[WARN] Failed to fetch API info: {any}\n", .{err});
        }
        
        // Fetch model card README
        if (self.fetchModelCard(hf_repo)) |readme| {
            defer self.allocator.free(readme);
            try self.extractSpecifications(&info, readme, hf_repo);
            try self.extractBenchmarks(&info, readme);
            try self.extractHardwareInfo(&info, readme);
        } else |err| {
            std.debug.print("[WARN] Failed to fetch model card: {any}\n", .{err});
        }
        
        // Determine categories and agent types
        try self.determineCategories(&info, hf_repo);
        try self.determineAgentTypes(&info);
        
        // Always include inference
        const inference = try self.allocator.dupe(u8, "inference");
        try info.agent_types.insert(0, inference);
        
        return info;
    }
    
    fn fetchAPIInfo(self: *HFModelCardExtractor, hf_repo: []const u8) ![]const u8 {
        _ = self;
        _ = hf_repo;
        // TODO: Re-implement with Zig 0.15.2 HTTP API
        return error.HTTPNotImplemented;
    }
    
    fn fetchModelCard(self: *HFModelCardExtractor, hf_repo: []const u8) ![]const u8 {
        _ = self;
        _ = hf_repo;
        // TODO: Re-implement with Zig 0.15.2 HTTP API
        return error.HTTPNotImplemented;
    }
    
    fn parseAPIInfo(self: *HFModelCardExtractor, info: *ModelInfo, json_data: []const u8) !void {
        const parsed = try std.json.parseFromSlice(
            std.json.Value,
            self.allocator,
            json_data,
            .{}
        );
        defer parsed.deinit();
        
        const root = parsed.value.object;
        
        // Extract key metadata
        if (root.get("downloads")) |downloads| {
            try info.hf_metadata.put("downloads", downloads);
        }
        if (root.get("likes")) |likes| {
            try info.hf_metadata.put("likes", likes);
        }
        if (root.get("pipeline_tag")) |pipeline| {
            try info.hf_metadata.put("pipeline_tag", pipeline);
        }
    }
    
    fn extractSpecifications(
        self: *HFModelCardExtractor,
        info: *ModelInfo,
        readme: []const u8,
        hf_repo: []const u8
    ) !void {
        // Extract parameter count from repo name
        if (std.mem.indexOf(u8, hf_repo, "B")) |_| {
            // Look for patterns like "7B", "1.5B"
            var it = std.mem.tokenizeAny(u8, hf_repo, "-_/");
            while (it.next()) |token| {
                if (std.mem.endsWith(u8, token, "B")) {
                    const key = try self.allocator.dupe(u8, "parameters");
                    const value = try self.allocator.dupe(u8, token);
                    try info.specifications.put(key, value);
                    break;
                }
            }
        }
        
        // Detect quantization
        if (std.mem.indexOf(u8, hf_repo, "GGUF")) |_| {
            const key = try self.allocator.dupe(u8, "quantization");
            var value: []const u8 = "unknown";
            
            if (std.mem.indexOf(u8, hf_repo, "Q4")) |_| {
                value = "Q4";
            } else if (std.mem.indexOf(u8, hf_repo, "Q8")) |_| {
                value = "Q8";
            } else if (std.mem.indexOf(u8, hf_repo, "Q6")) |_| {
                value = "Q6";
            }
            
            const value_copy = try self.allocator.dupe(u8, value);
            try info.specifications.put(key, value_copy);
        }
        
        _ = readme; // TODO: Parse README for more specs
    }
    
    fn extractBenchmarks(self: *HFModelCardExtractor, info: *ModelInfo, readme: []const u8) !void {
        // Simple regex-like search for benchmark patterns
        const benchmarks = [_]struct { name: []const u8, pattern: []const u8 }{
            .{ .name = "gsm8k", .pattern = "GSM8K" },
            .{ .name = "humaneval", .pattern = "HumanEval" },
            .{ .name = "mbpp", .pattern = "MBPP" },
            .{ .name = "mmlu", .pattern = "MMLU" },
        };
        
        for (benchmarks) |bench| {
            if (std.mem.indexOf(u8, readme, bench.pattern)) |idx| {
                // Look for a number after the benchmark name
                const search_start = idx + bench.pattern.len;
                if (search_start < readme.len) {
                    const remaining = readme[search_start..];
                    
                    // Simple number extraction
                    var score_str = std.ArrayList(u8){};
                    defer score_str.deinit();
                    
                    for (remaining) |c| {
                        if (std.ascii.isDigit(c) or c == '.') {
                            try score_str.append(c);
                        } else if (score_str.items.len > 0) {
                            break;
                        }
                    }
                    
                    if (score_str.items.len > 0) {
                        if (std.fmt.parseFloat(f64, score_str.items)) |score| {
                            const name_copy = try self.allocator.dupe(u8, bench.name);
                            const date_copy = try self.allocator.dupe(u8, "2026-01");
                            try info.benchmarks.put(name_copy, .{
                                .score = score,
                                .date = date_copy,
                            });
                            self.log("Found {s}: {d:.2}", .{ bench.name, score });
                        } else |_| {}
                    }
                }
            }
        }
    }
    
    fn extractHardwareInfo(self: *HFModelCardExtractor, info: *ModelInfo, readme: []const u8) !void {
        // Look for GPU memory patterns
        if (std.mem.indexOf(u8, readme, "GB")) |_| {
            // Simple heuristic: look for patterns like "16GB GPU" or "requires 8GB"
            var it = std.mem.tokenizeAny(u8, readme, " \t\n");
            while (it.next()) |token| {
                if (std.mem.endsWith(u8, token, "GB")) {
                    const key = try self.allocator.dupe(u8, "min_gpu_memory");
                    const value = try self.allocator.dupe(u8, token);
                    try info.hardware.put(key, value);
                    break;
                }
            }
        }
    }
    
    fn determineCategories(self: *HFModelCardExtractor, info: *ModelInfo, hf_repo: []const u8) !void {
        var categories = std.StringHashMap(void).init(self.allocator);
        defer categories.deinit();
        
        // Check repo name patterns
        var it = NAME_PATTERNS.iterator();
        while (it.next()) |entry| {
            const category = entry.key;
            const patterns = entry.value;
            
            for (patterns) |pattern| {
                if (std.mem.indexOf(u8, hf_repo, pattern)) |_| {
                    try categories.put(category, {});
                    break;
                }
            }
        }
        
        // Default to code if nothing found
        if (categories.count() == 0) {
            try categories.put("code", {});
        }
        
        // Convert to array
        var cat_iter = categories.iterator();
        while (cat_iter.next()) |entry| {
            const cat_copy = try self.allocator.dupe(u8, entry.key_ptr.*);
            try info.orchestration_categories.append(cat_copy);
        }
    }
    
    fn determineAgentTypes(self: *HFModelCardExtractor, info: *ModelInfo) !void {
        _ = self;
        _ = info;
        // Agent types determined by categories
        // Always include "inference" which is added in extractModelInfo
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    
    if (args.len < 2) {
        const stderr = std.fs.File.stderr();
        try stderr.writeAll("Usage: ");
        try stderr.writeAll(args[0]);
        try stderr.writeAll(" <registry_path> [--test HF_REPO] [--verbose]\n");
        std.process.exit(1);
    }
    
    var verbose = false;
    var test_repo: ?[]const u8 = null;
    
    // Parse options
    for (args[1..]) |arg| {
        if (std.mem.eql(u8, arg, "--verbose") or std.mem.eql(u8, arg, "-v")) {
            verbose = true;
        } else if (std.mem.eql(u8, arg, "--test")) {
            // Next arg should be repo
            continue;
        } else if (test_repo == null and args.len > 2) {
            test_repo = arg;
        }
    }
    
    var extractor = HFModelCardExtractor.init(allocator, verbose);
    defer extractor.deinit();
    
    if (test_repo) |repo| {
        std.debug.print("\n{'=':**<60}\n", .{});
        std.debug.print("Testing model: {s}\n", .{repo});
        std.debug.print("{'=':**<60}\n\n", .{});
        
        var info = try extractor.extractModelInfo(repo);
        defer info.deinit();
        
        std.debug.print("Categories: ", .{});
        for (info.orchestration_categories.items) |cat| {
            std.debug.print("{s} ", .{cat});
        }
        std.debug.print("\n", .{});
        
        std.debug.print("Benchmarks: {d} found\n", .{info.benchmarks.count()});
        var bench_iter = info.benchmarks.iterator();
        while (bench_iter.next()) |entry| {
            std.debug.print("  {s}: {d:.2}\n", .{ entry.key_ptr.*, entry.value_ptr.score });
        }
    } else {
        std.debug.print("Registry enrichment not yet implemented\n", .{});
        std.debug.print("Use --test HF_REPO to test single model extraction\n", .{});
    }
}
